Testing/C2DTests.cs

```diff
@@ -1,97 +1,144 @@
         }
 
         [Test]
         public async Task SendD2CWithRetainFlagTrue()
         {
             // Arrange
-            await device.ConnectDevice();
-
-            var retainFlag = true;
+            var receiver = new Receiver(eventHubCompatibleEndpoint, eventHubName, iotHubSasKey);
+            var device = new Device(iotHubDeviceConnectionString);
             var payload = Guid.NewGuid().ToString();
 
+            await device.ConnectDevice();
+
             // Act
-            await device.SendDeviceToCloudMessageAsync(payload, retainFlag);
+            await device.SendDeviceToCloudMessageAsync(payload, true);
 
-            var cancellationSource = new CancellationTokenSource();
-            cancellationSource.CancelAfter(3000);
-            var messages = await receiverConsumer.ReceiveMessagesFromDeviceAsync(cancellationSource.Token);
+            var cancellationSource = new CancellationTokenSource(3000);
+            var messages = await receiver.ReceiveMessagesFromDeviceAsync(cancellationSource.Token);
 
             // Assert - verify message was received + mqtt-retain set to true
             var sentMessage = messages.FirstOrDefault(x => x.Payload == payload);
             Assert.IsTrue(sentMessage != null && sentMessage.RetainFlag == "true");
         }
 
         [Test]
+        public async Task ReceiveAllC2DMessagesWhileDisconnected()
+        {
+            // Arrange
+            var device = new Device(iotHubDeviceConnectionString);
+            var sender = new Sender(iotHubConnectionString, deviceId);
+            var firstPayload = Guid.NewGuid().ToString();
+            var secondPayload = Guid.NewGuid().ToString();
+            var payloads = new ConcurrentBag<string>();
+            Action<MqttApplicationMessageReceivedEventArgs> applicationMessageReceived = (MqttApplicationMessageReceivedEventArgs e) =>
+            {
+                payloads.Add(e.ApplicationMessage.ConvertPayloadToString());
+            };
+
+            // Act
+            await device.ConnectDevice();
+            await device.SubscribeToEventAsync(applicationMessageReceived);
+            await sender.SendCloudToDeviceMessageAsync(firstPayload);
+            Assert.IsTrue(RetryUntilSuccessOrTimeout(() => payloads.FirstOrDefault(x => x == firstPayload) != null, TimeSpan.FromSeconds(10)));
+
+            await device.DisconnectDevice();
+            Thread.Sleep(15000);
+            await sender.SendCloudToDeviceMessageAsync(secondPayload);
+
+            device = new Device(iotHubDeviceConnectionString);
+            await device.ConnectDevice();
+            await device.SubscribeToEventAsync(applicationMessageReceived);
+
+            // Assert
+            Assert.IsTrue(RetryUntilSuccessOrTimeout(() => payloads.FirstOrDefault(x => x == secondPayload) != null, TimeSpan.FromSeconds(10)));
+        }
+
```

Comment:
Please add the [Test] here as well

Reply:
Ha, that's funny. Thank you :) Great catch!